### ISSUE TITLE
When forked, "New Branch" is confusing for students

### ISSUE THREAD

**Issue by charles-allen** [desktop/desktop#12534]
## Describe the feature or problem you’d like to solve

_Two related problems:_

- **As a** University Lecturer  
  **I want** students to make weekly branches and merges within a student fork
  **So that** students can learn a standard development process, I can review their code frequently, they can begin from a starter template, they avoid fiddly repo creation or fork decoupling actions, they are able to diff against my solution code.  
**=> problem: re-branch is not possible in GHD**

- **As a** Campus Advisor  
**I want** to promote a simple & intuitive Git Client for teaching  
**So that** more teachers are able to integrate source control, code review, CICD, etc into their classes  
**=> problem: I feel uncomfortable promoting GHD when it cannot do something fundamental that every other client in the industry can**

## Proposed solution

- **Solution:** Let the user change the source for new branches (i.e. "change" link next to the current source).
- **Benefits:** Without this feature, my students are forced to leave GitHub Desktop to use the CLI or another Git Client. At that point there is no point using GitHub Desktop. IMO this is a deal-breaker preventing adoption of GHD. With this feature, GHD is my first choice for teaching.

### My Use Case

I am a lecturer and our university's GitHub Campus Advisor. I'm trying to promote using GitHub in the classroom. I want to use industry tools & standard development processes as much as possible because I think the best learning environment is a real environment. I don't want to teach anything artificial that our students have to unlearn later.

I picked GitHub Desktop because it's no fuss and it "just works". It seems to me that education should be a major market for GitHub Desktop for this reason.

My students fork their repos from my template. **I want to use a fork so that they can easily diff against my solution code.** Each week they create a branch and PR their work (to their fork, not upstream). This process teaches them a normal development workflow, facilitates feedback, and allows them to compare their code to mine.

_Note: When cloning, my students explicitly select an option to make PRs against their forks, not the upstream._

I accept that there are almost certainly work-arounds; but this does not address problem (2) -- **in industry the expectation is that I'm able to branch; I don't want to break that in my classroom environment.**

### This is not about UX (default source)

IMO defaulting to the upstream is incredibly unintuitive, especially when my forked `develop` & `origin/develop` have commits that aren't in `upstream/develop`. I would have preferred a warning dialog "your branch is out of date", and perhaps a checkbox "always pull before branching". However, this is besides the point!

**I don't care what the default is. I just want to be able to change it, without leaving GHD.**

### Past Discussions

There is some existing discussion here: https://github.com/desktop/desktop/issues/9384. I tried to pick out some key quotes:

> This was an intentional change in #7762 based on the vast majority of forks being used to contribute to the upstream branch. So the previous behavior often would mean people working from forks in Desktop were starting new branches way out of date from the latest commits, and this was intended to help people ensure they're up to date when starting on a new branch.

> The reason why we've chosen not to provide a toggle in this particular case is that we think it's such a limited edge case

**I am proposing education as a not-so-limited edge-case. But I accept that other universities may have different processes or be more open to work-arounds (I have a particular aversion to diverging from industry), so I don't know how big this part of the market is.**

**Comment by tidy-dev** [comment]
@charles-allen Thank you for your issue! If I am understanding correctly, your main concern is that when you create a fork by default and then you create a new branch, it will be based off of the original repo (upstream). You want it based off the default branch of your fork. There is a repository setting to allow you to toggle this. Closing this issue but of course comment back if this is not what you were looking for.

Top app menu > Repository Settings > Fork Behavior
![image](https://user-images.githubusercontent.com/75402236/123793692-01255680-d8b0-11eb-984d-5a012ebe8221.png)

**Comment by charles-allen** [comment]
**TL;DR: default values are a convenience you can hide in prefs. But silently obstructing the choice during the primary UI flow for doing it is a bad experience for the user. Either don't be silent or don't obstruct the choice.**

-----

@tiny-dev - My apologies. I definitely 100% selected that when I originally cloned! However it seems to be set to wrong value now. Perhaps I deleted the repo and recloned outside GHD. Perhaps I just removed it from GHD and then re-added it. I will experiment to familiarize myself with such cases.

**This absolutely does solve my first problem**; being aware of it really helps me -- thank you! I can make this work in my classes.

## Sharp Corners

At first I thought it solved my second problem too, but after playing around with it more, it turns out that this actually is about UX - not the which source is default, but explaining why it's locked & how to change it. I still don't feel comfortable pushing this on other lecturers:
- I guarantee some proportion of students will pick the wrong option
- It's non-obvious that the setting is "wrong"
- It's not clear that this even is a choice
- It's too non-intuitive to expect your users to search through GHD-specific prefs for a setting that isn't even mentioned on the new branch UI
- it's further confused by the fact I'm denied any choice on develop, but as soon as I'm on a branch, I'm presented with a beautiful UI to select between sources

Most students & some lecturers won't notice when this setting is wrong: 

| A | B |
| ------------- | ------------- |
| <img width="416" alt="Screen Shot 2021-06-30 at 06 09 39" src="https://user-images.githubusercontent.com/31621095/123879009-5e7dde00-d96a-11eb-9533-b3f4b3a8a4ff.png"> | <img width="416" alt="Screen Shot 2021-06-30 at 06 10 08" src="https://user-images.githubusercontent.com/31621095/123879017-6178ce80-d96a-11eb-86af-ffd0a0116454.png"> |

_Edit: For context on where I'm coming from; when you're teaching a room full of completely new devs, any unnecessary complexity is your enemy. Even the simplest instructions will trip a fraction of these new devs. It's common for them to be derailed by trivial errors like typos, bad capitalization, not noticing that a CLI command failed, etc. In this environment, every edge case that I have to explain is undesirable. I noticed that I'd branched from the wrong fork. My students would silently continue with their repo in the wrong state. I can look out for some issues like this in my classes. But each one has a toll on learning, and I can't look out for issues in other lecturer's classes._

## Solutions

IMO, this needs to be more intuitive & frictionless. Either:
- the new branch UI should clearly indicate why the source branch was locked, and provide either an inline action to change the setting directly, or a link to open repo fork prefs, ex: "based on [your fork settings](link_to_fork_prefs)" or "[change your fork settings](link_to_fork_prefs)".
- or my preferred solution -- just **use your beautiful choice UI when develop is checked out => offer develop AND upstream/develop** (you can still default based on fork mode) -- this probably implies that you should have 3 choices when a branch is checked out (again with the default based on fork mode)

| You have this | ...so why not this? |
| ------------- | ------------- |
| <img width="416" alt="Screen Shot 2021-06-30 at 06 10 25" src="https://user-images.githubusercontent.com/31621095/123879492-317dfb00-d96b-11eb-8758-2bdd3a81f109.png"> | <img width="423" alt="Screen Shot 2021-06-30 at 06 12 38" src="https://user-images.githubusercontent.com/31621095/123879497-3478eb80-d96b-11eb-8296-78f7cc6b554a.png"> |

**Comment by charles-allen** [comment]
Sorry for all the edits & long posts!

I realize this is probably a minor concern for the team and you're all very busy doing more important things. I appreciate your time helping me, and I don't want to seem ungrateful. Just trying to give you the best data I can to help make GHD the best product for me :D

**Comment by tidy-dev** [comment]
@charles-allen The thorough posts are great for us to get a solid understanding of your use case. I am going to open this back and mark as an enhancement for future consideration. Unfortunately, as you read in some of the other issues, we have not found this use case to be a common one and not inline with our current roadmap so I am not sure when it may actually be considered further. Tho, I do really like the suggestions to make this more clear especially for new devs. Thank you for your time in laying these suggestions.

**Comment by charles-allen** [comment]
Just fighting for my corner ;) I'm not really sure what the target audience is for Github Desktop, but in most respects it really is ideal for teaching. I think I needed this much detail to really get into the trouble a little friction is when you have a large class of students.

If you need any more input or feedback, please feel free to `@` ping me.

If you don't mind, I will also update the title to better reflect where the discussion has reached.

**Comment by billygriffin** [comment]
Hi @charles-allen, thanks so much for explaining your concerns so well! Of your suggestions, I think the second one might actually create *more* confusion. We try to make things simpler where possible, and for people who set things up correctly when asked about how they're using their fork, asking every single time they want to create a branch feels like it'll add confusion for the happy path case.

That said, I do think providing a link to where they can update their fork preferences in that explainer text is a worthwhile improvement imo and something we've done in other places in the app. If someone wants to take a stab at that in a PR, we'd be happy to review that and see how it feels.

**Comment by charles-allen** [comment]
> I think the second one might actually create _more_ confusion [...] for people who set things up correctly when asked about how they're using their fork

This makes sense. I was quite focused on my corner, but now that the issue has been narrowed in scope & value, it's hard to justify going against the many; and this 2nd approach contradicts the entire design decision to have a settings "switch" for forks.

> something we've done in other places in the app. If someone wants to take a stab at that in a PR, we'd be happy to review that and see how it feels.

Can you point me at an similar example in the code? If so, I'm happy to have a go at a PR (though I can't promise I'll succeed).

**Comment by billygriffin** [comment]
@charles-allen This is the place I was thinking of.

<img width="515" alt="Screen Shot 2021-07-16 at 9 56 53 AM" src="https://user-images.githubusercontent.com/5091167/125975558-14ac3c75-7ef3-45ed-b5a6-5aa33c480ea1.png">

I'm about to hop into a meeting so I can't go look for it in the code but I think you should be able to search for the relevant verbiage and find it hopefully without too much trouble. The part that's most relevant to this use case is `You can also choose an email local to this repository from the repository settings` where repository settings is linked and takes you to the right place.

**Comment by charles-allen** [comment]
> I think you should be able to search for the relevant verbiage and find it hopefully without too much trouble. The part that's most relevant to this use case is `You can also choose an email local to this repository from the repository settings` where repository settings is linked and takes you to the right place.

Thanks! Yes, I should be able to find that; and it will do me good to explore a bit. Wish me luck :D

**Comment by geophilusd** [comment]
@charles-allen Are you working on this one? I am interested in giving this a try.

**Comment by charles-allen** [comment]
> @charles-allen Are you working on this one? I am interested in giving this a try.

Go for it!

I haven't started, and I have a few of these jobs waiting for a moment of peace that may never arrive.
